### PR TITLE
Issue #348: Remove composer runtime dependency.

### LINKED ISSUE #348: Get rid of runtime composer dependency
* No composer install requirement during installation
* No runtime  composer autoloader usage
* modman dependency on composer classes
* Replace guzzle with the `\Mage_HTTP_Client` factory and `\Mage_HTTP_IClient` (or just plain PHP)

### PR DESCRIPTION
**Pull request by Vinai**
Replace Guzzle with simple PHP implementation and remove runtime composer dependency.

Closes #348

### CODE REVIEW

File: lib/LizardsAndPumpkins/src/Api/PhpStreamHttpApiClient.php [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+    {
+        $httpRequestContext = stream_context_create(['http' => [
+            'method' => 'PUT',
+            'header' => $this->concatHeaderNamesAndValues($headers),
+            'content' => $body,
+        ]]);
+        return $this->doHttpRequest($url, $httpRequestContext);
+    }
+
+    public function getRequest(string $url, array $headers): string
```
Comment (Schrank) [inline lib/LizardsAndPumpkins/src/Api/PhpStreamHttpApiClient.php]: The name getRequest is  ambiguous in context of `putRequest` vs `getRequest` and `getRequest` vs `getResponse`

File: lib/LizardsAndPumpkins/src/Api/PhpStreamHttpApiClient.php [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+            'method' => 'GET',
+            'header' => $this->concatHeaderNamesAndValues($headers),
+        ]]);
+        return $this->doHttpRequest($url, $httpRequestContext);
+    }
+
+    private function doHttpRequest(string $url, $httpRequestContext): string
+    {
+        $this->validateUrl($url);
+        $http_response_header = null;
```
Comment (Schrank) [inline lib/LizardsAndPumpkins/src/Api/PhpStreamHttpApiClient.php]: either camelCase or underline, see `$responseBody `

File: lib/LizardsAndPumpkins/src/Api/PhpStreamHttpApiClient.php [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+        if ($parts['scheme'] !== 'http' && $parts['scheme'] !== 'https') {
+            $message = sprintf('Lizards & Pumpkins API URL must start with "http" or "https", got "%s".', $url);
+            throw new InvalidUrlException($message);
+        }
+    }
+
+    private function parseResponseStatusCode(array $httpResponseHeaders): int
+    {
+        foreach ($httpResponseHeaders as $header) {
+            if (preg_match('#HTTP/\S+ (?<responseCode>\d+)#', $header, $matches)) {
```
Comment (Schrank) [inline lib/LizardsAndPumpkins/src/Api/PhpStreamHttpApiClient.php]: Did you think about just using `substr()` I don't know how strict the standard is. substr feels fragile to me, an idea.

File: lib/LizardsAndPumpkins/src/Api/PhpStreamHttpApiClient.php [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+                return (int) $matches['responseCode'];
+            }
+        }
+
+        return 0;
+    }
+
+    private function validateHttpResponse(array $http_response_header = null)
```
Comment (Schrank) [inline lib/LizardsAndPumpkins/src/Api/PhpStreamHttpApiClient.php]: parameter in camel case

File: lib/LizardsAndPumpkins/src/Api/PhpStreamHttpApiClient.php [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+            );
+            throw new RequestFailedException($message);
+        }
+    }
+
+    /**
+     * @param string[] $headerNameToValueMap
+     * @return string[]
+     */
+    private function concatHeaderNamesAndValues(array $headerNameToValueMap): array
```
Comment (Schrank) [inline lib/LizardsAndPumpkins/src/Api/PhpStreamHttpApiClient.php]: If we use header names in an array, we can only have each header once. I don't think this is a problem.

Magento is NOT replacing headers by default (not sure why)

File: tests/Unit/Suites/Api/PhpStreamHttpApiClientTest.php [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+
+    public function testThrowsAnExceptionOnNonSuccessfulGetRequests()
+    {
+        $this->expectException(RequestFailedException::class);
+        $expectedMessage = 'The HTTP response status code of the Lizards & Pumpkins API ' .
+                   'is not within the expected 200-299 range, got 403';
+        $this->expectExceptionMessage($expectedMessage);
+        self::$simulateHttpResponseBody = '';
+        
+        $failingTestApiClient = $this->createTestApiClientWithResponseCode('HTTP/1.1 403 Forbidden');
```
Comment (Schrank) [inline tests/Unit/Suites/Api/PhpStreamHttpApiClientTest.php]: The excessive(?) tests for all the http status codes, helped me to identify a bug with the >200 <=300 in the last implementation. Maybe worth adding them?

File: lib/LizardsAndPumpkins/src/Api/PhpStreamHttpApiClient.php (hunk 2 above)
Comment (Vinai) [inline lib/LizardsAndPumpkins/src/Api/PhpStreamHttpApiClient.php]: This is a magic variable name that gets set in local scope when php stream wrappers are used. I can not chage it without a BC breaking PR to the PHP core.

File: lib/LizardsAndPumpkins/src/Api/PhpStreamHttpApiClient.php (hunk 3 above)
Comment (Vinai) [inline lib/LizardsAndPumpkins/src/Api/PhpStreamHttpApiClient.php]: I originally used `substr` but with case change and matching  a range of HTTP versions it was a lot more complex.

File: lib/LizardsAndPumpkins/src/Api/PhpStreamHttpApiClient.php (hunk 4 above)
Comment (Vinai) [inline lib/LizardsAndPumpkins/src/Api/PhpStreamHttpApiClient.php]: I want to communicate that it's the magic variable defined by PHP that is passed as an argument.

File: lib/LizardsAndPumpkins/src/Api/PhpStreamHttpApiClient.php (hunk 5 above)
Comment (Vinai) [inline lib/LizardsAndPumpkins/src/Api/PhpStreamHttpApiClient.php]: It is legal to define multiple headers. I chose this format to not break the contract you defined.

File: lib/LizardsAndPumpkins/src/Api/PhpStreamHttpApiClient.php (hunk 1 above)
Comment (Vinai) [inline lib/LizardsAndPumpkins/src/Api/PhpStreamHttpApiClient.php]: Since this is a BC break to the interface you defined I think it's better to move that into another PR.

File: tests/Unit/Suites/Api/PhpStreamHttpApiClientTest.php (hunk 6 above)
Comment (Vinai) [inline tests/Unit/Suites/Api/PhpStreamHttpApiClientTest.php]: Good point, I'll add edge case tests.

File: lib/LizardsAndPumpkins/src/Api/PhpStreamHttpApiClient.php (hunk 4 above)
Comment (Schrank) [inline lib/LizardsAndPumpkins/src/Api/PhpStreamHttpApiClient.php]: I'm not familiar with this variable, so maybe others - rename it to `magic_http_response_header`? Or `php_defined_http_response_header`? (I don't like the long name)

Comment (tim-bezhashvyly) [inline lib/LizardsAndPumpkins/src/Api/PhpStreamHttpApiClient.php]: Missing PHPDoc.

File: lib/LizardsAndPumpkins/src/Api/PhpStreamHttpApiClient.php [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+            throw new InvalidUrlException(sprintf('Unable to parse host from Lizards & Pumpkins API URL "%s".', $url));
+        }
+
+        if ($parts['scheme'] !== 'http' && $parts['scheme'] !== 'https') {
+            $message = sprintf('Lizards & Pumpkins API URL must start with "http" or "https", got "%s".', $url);
+            throw new InvalidUrlException($message);
+        }
+    }
+
+    private function parseResponseStatusCode(array $httpResponseHeaders): int
```
Comment (tim-bezhashvyly) [inline lib/LizardsAndPumpkins/src/Api/PhpStreamHttpApiClient.php]: Missing PHPDoc.

File: lib/LizardsAndPumpkins/src/Api/PhpStreamHttpApiClient.php [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+        }
+    }
+
+    /**
+     * @param string[] $headerNameToValueMap
+     * @return string[]
+     */
+    private function concatHeaderNamesAndValues(array $headerNameToValueMap): array
+    {
+        return array_map(function ($headerName, $headerValue) {
```
Comment (tim-bezhashvyly) [inline lib/LizardsAndPumpkins/src/Api/PhpStreamHttpApiClient.php]: Would it make sense to type-hint closure arguments?

File: lib/LizardsAndPumpkins/src/Api/PhpStreamHttpApiClient.php [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+            return $headerName . ': ' . $headerValue;
+        }, array_keys($headerNameToValueMap), array_values($headerNameToValueMap));
```
Comment (tim-bezhashvyly) [inline lib/LizardsAndPumpkins/src/Api/PhpStreamHttpApiClient.php]: This is risky as PHP doesn't guarantee an order of elements. I normally just use `array_keys` and pass values inside of a closure `use` operator.

File: magento-plugin/app/code/community/LizardsAndPumpkins/MagentoConnector/Model/Observer/CmsBlockSaveCommitAfter.php [hunk 10]
```diff
@@ -1,7 +1,6 @@
 <?php
-declare(strict_types=1);
 
-require_once __DIR__ . '/../../../../../../../../vendor/autoload.php';
+declare(strict_types=1);
```
Comment (tim-bezhashvyly) [inline magento-plugin/app/code/community/LizardsAndPumpkins/MagentoConnector/Model/Observer/CmsBlockSaveCommitAfter.php]: I guess we agreed to have spaces around equation.

File: lib/LizardsAndPumpkins/src/Api/PhpStreamHttpApiClient.php (hunk 4 above)
Comment (Vinai) [inline lib/LizardsAndPumpkins/src/Api/PhpStreamHttpApiClient.php]: Renamed into camel case, it's not that important there.

File: lib/LizardsAndPumpkins/src/Api/HttpApiClient.php [hunk 11]
```diff
@@ -6,8 +6,30 @@
...
+     * Headers array format:
+     * [
+     *    'Header-Name' => 'Header Value'
+     * ]
+     * 
+     * @param string $url
+     * @param string $body
+     * @param string[] $headers
+     * @return string
+     * @throws \LizardsAndPumpkins\MagentoConnector\Api\InvalidUrlException
```
Comment (Vinai) [inline lib/LizardsAndPumpkins/src/Api/HttpApiClient.php]: It's part of the inteface, there it makes sense.

File: lib/LizardsAndPumpkins/src/Api/HttpApiClient.php [hunk 12]
```diff
@@ -6,8 +6,30 @@
...
+    /**
+     * Headers array format:
+     * [
+     *    'Header-Name' => 'Header Value'
+     * ]
+     * 
+     * @param string $url
+     * @param string[] $headers
+     * @return string
+     * @throws \LizardsAndPumpkins\MagentoConnector\Api\InvalidUrlException
```
Comment (Vinai) [inline lib/LizardsAndPumpkins/src/Api/HttpApiClient.php]: As part of the interface it makes sense to have it.

Comment (tim-bezhashvyly) [inline lib/LizardsAndPumpkins/src/Api/HttpApiClient.php]: For IDE or developer?

Comment (Vinai) [inline lib/LizardsAndPumpkins/src/Api/HttpApiClient.php]: Developer

File: lib/LizardsAndPumpkins/src/Api/PhpStreamHttpApiClient.php [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+    public function getRequest(string $url, array $headers): string
+    {
+        $httpRequestContext = stream_context_create(['http' => [
+            'method' => 'GET',
+            'header' => $this->concatHeaderNamesAndValues($headers),
+        ]]);
+        return $this->doHttpRequest($url, $httpRequestContext);
+    }
+
+    private function doHttpRequest(string $url, $httpRequestContext): string
```
Comment (Vinai) [inline lib/LizardsAndPumpkins/src/Api/PhpStreamHttpApiClient.php]: Hm, what should it be? Resource?

File: lib/LizardsAndPumpkins/src/Api/PhpStreamHttpApiClient.php (hunk 8 above)
Comment (Vinai) [inline lib/LizardsAndPumpkins/src/Api/PhpStreamHttpApiClient.php]: Done.

File: lib/LizardsAndPumpkins/src/Api/PhpStreamHttpApiClient.php (hunk 9 above)
Comment (Vinai) [inline lib/LizardsAndPumpkins/src/Api/PhpStreamHttpApiClient.php]: Really? I've never experienced that. Actually, PHP does guarantee ordering of arrays afaik.

File: magento-plugin/app/code/community/LizardsAndPumpkins/MagentoConnector/Model/Observer/CmsBlockSaveCommitAfter.php (hunk 10 above)
Comment (Vinai) [inline magento-plugin/app/code/community/LizardsAndPumpkins/MagentoConnector/Model/Observer/CmsBlockSaveCommitAfter.php]: I think we agreed that it doesn't matter until PHPStorm supports a separate configuration in the stable version.

File: lib/LizardsAndPumpkins/src/Api/PhpStreamHttpApiClient.php (hunk 9 above)
Comment (Schrank) [inline lib/LizardsAndPumpkins/src/Api/PhpStreamHttpApiClient.php]: Me neither, but @tim-bezhashvyly is right, in the doc there is not a single word about specification of order http://php.net/manual/en/function.array-values.php neither any connection between array_values and array_keys.

BUT everyone is expecting that and the implemention (feels) looks like it is using foreach() internally to generate these arrays.

If this behaviour breaks in PHP we'll get a lot of problems with a lot of frameworks.
